IPIP/0373-double-hash-dht.md

```diff
@@ -0,0 +1,404 @@
+5. DHT servers compute `Hash(MH)`.
+6. DHT servers find the 20 closest peers to `Hash(MH)` in XOR distance in their Routing Table.
+7. DHT servers return the 20 `PeerID`s and `multiaddrs` of these peers to Client.
+8. Client sends a DHT lookup request for `CID` to the closest peers in XOR distance to `Hash(MH)` that it received.
+9. Go to 4.
+10. The DHT servers storing the Provider Record(s) associated with `MH` send them to Client. (Currently, if a Provider Record has been published less than 30 min before being requested, the DHT servers also send the `multiaddresses` of the Content Provider to Client).
+11. If the response from the DHT server doesn't include the `multiaddrs` associated with the Content Providers' `PeerID`s, Client performs a DHT `FindPeer` request to find the `multiaddrs` of the returned `PeerID`s.
+12. Client sends a Bitswap request for `CID` to the Content Provider (known `PeerID` and `multiaddrs`).
+13. Content Provider sends the requested content back to Client.
+
+### Double Hash DHT design
+
+**Publish Process**
+1. Content Provider wants to publish some content with identifier `CID`.
+2. Content Provider computes `HASH2 = SHA256(bytes("CR_DOUBLEHASH") || MH)` (`MH` is the MultiHash included in the CID).
+3. Content Provider starts a DHT lookup request for the 20 closest `PeerID`s in XOR distance to `HASH2`.
+4. Content Provider encrypts its own `PeerID` (`CPPeerID`) with `MH`, using AES-GCM. `EncPeerID = [0x8040, Nonce, payload_len, AESGCM(MH, Nonce, CPPeerID)]`
+5. Content Provider takes the current timestamp `TS`.
+6. Content Provider signs `EncPeerID` and `TS` using its private key. `Signature = Sign(privkey, EncPeerID || TS)`
+7. Content Provider computes `ServerKey = SHA256(bytes("CR_SERVERKEY") || MH)`.
+8. Once the lookup request has returned the 20 closest peers, Content Provider sends a Publish request to these DHT servers. The Publish request contains [`HASH2`, `EncPeerID`, `TS`, `Signature`, `ServerKey`].
+9. Each DHT server verifies `Signature` against the `PeerID` of the Content Provider used to open the libp2p connection. `Verify(CPPeerID, Signature, EncPeerID || TS)`. It verifies that `TS` is younger than `48h` and isn't in the future. If invalid, send an error to the client. <!-- TODO: define error -->
+10. Each DHT server adds an entry in their Provider Store for `HASH2` -> `ServerKey` -> `CPPeerID` -> [`EncPeerID`, `TS`, `Signature`], with `CPPeerID` being the `PeerID` of the Content Provider (see [provider store](#provider-store)). If there is already an entry including `CPPeerID` for `HASH2` -> `ServerKey`, and if the `TS` of the new valid entry is newer than the existing `TS`, overwrite the entry in the Provider Store. Else drop the new entry.
+11. Each DHT server confirms to Content Provider that the Provider Record has been successfully added.
+12. The proces is over once Content Provider has received 20 confirmations.
```

Comment:
Nit: `Confirmation` sounds to me that the remote peer needs to acknowledge the receipt of the provider record. So this spec requires 20 successful RPC's. However, in the current implementation a significant fraction of the `ADD_PROVIDER` RPC's error out and we basically just live with it.

The previous steps set the stage where all DHT servers are well functioning so this step 12 may be fine here. However, what do you think of a brief discussion about how to handle errors in this final step?

Reply:
The number of Provide ACK isn't modified by this Spec, it is expected to stay the same as it currently is. There isn't a proper Spec for the DHT yet, so I couldn't add a reference. Do you know what is the current state of the implementation (regarding the number of Provide ACKs that are required) ?

 The only difference in the Provide Operation is the content that is transmitted from the Content Provider to the DHT Servers.

Reply:
There is no required number of ACKs :/ In the current implementation we try to find the 20 closest peers and as soon as we have them we fire 20 concurrent `ADD_PROVIDER` RPCs. If any call fails we just [log the error](https://github.com/libp2p/go-libp2p-kad-dht/blob/76b0f760072c249de070ee285c434064f50bdcc1/routing.go#L430-L434
) and call it day 🙈

Reply:
This sounds like something to modify in the Provide process directly 👍🏻